README.md
# Stable Diffusion Web UI Docker for Intel Arc GPUs

<a href="https://hub.docker.com/r/nuullll/ipex-arc-sd">
  <img src="https://img.shields.io/docker/pulls/nuullll/ipex-arc-sd" />
</a>

- [Documentation](https://blog.nuullll.com/ipex-sd-docker-for-arc-gpu/#/)
- [Getting Started](https://blog.nuullll.com/ipex-sd-docker-for-arc-gpu/#/getting-started)
- [FAQ](https://blog.nuullll.com/ipex-sd-docker-for-arc-gpu/#/faq)
- [Release Notes](https://blog.nuullll.com/ipex-sd-docker-for-arc-gpu/#/release-notes)

The [docker image](https://hub.docker.com/r/nuullll/ipex-arc-sd) includes
- Intel oneAPI DPC++ runtime libs _(Note: compiler executables are not included)_
- Intel oneAPI MKL runtime libs
- Intel oneAPI compiler common tool `sycl-ls`
- Intel Graphics driver
- Basic python environment

The Stable Diffusion Web UI variant used by the image: [SD.Next](https://github.com/vladmandic/automatic)

- Intel Extension for Pytorch (IPEX) and other python packages will be installed by [SD.Next](https://github.com/vladmandic/automatic) dynamically

## (For Developers) Build docker image locally

```powershell
docker build -t ipex-arc-sd -f Dockerfile .
```

Refer to [Dockerfile](./Dockerfile) for available build arguments.

## Contributors

<a href="https://github.com/Nuullll/ipex-sd-docker-for-arc-gpu/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=Nuullll/ipex-sd-docker-for-arc-gpu" />
</a>


docs/release-notes.md
# Release Notes

## [v0.6](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.6/images/sha256-30bdf186bc21abbcbb1d59ee87b4a726af9aa93794543121caf58ba95f44caaa?context=explore) (latest)

Compressed image size: 845.44 MB

Major changes:

- Uplift oneAPI to 2023.2 to support IPEX XPU 2.0

The [image](https://hub.docker.com/r/nuullll/ipex-arc-sd) (Ubuntu 22.04 based) includes:

- Intel oneAPI DPC++ runtime libs (2023.2.1) (Note: compiler executables are not included)
- Intel oneAPI MKL runtime libs (2023.2.0)
- Intel oneAPI compiler common tool sycl-ls (2023.2.1)
- Intel Graphics driver (1.3.26241.33-647~22.04)
- Basic python environment (3.10.6)

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4502)

## [v0.5](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.5/images/sha256-bb556a04a3ad6d331582ad1d64e79a123650fd43981d2bdd3c2e1f639bde818c?context=explore)

Compressed image size: 831.44 MB

Major changes:

- Allow extension installation by default (`--insecure`)
- Skip startup git operations by default (`--skip-git`)
- Use faster offline git repo check

The [image](https://hub.docker.com/r/nuullll/ipex-arc-sd) (Ubuntu 22.04 based) includes:

- Intel oneAPI DPC++ runtime libs (2023.1) (Note: compiler executables are not included)
- Intel oneAPI MKL runtime libs (2023.1)
- Intel oneAPI compiler common tool sycl-ls (2023.1)
- Intel Graphics driver (1.3.26241.21-647~22.04)
- Basic python environment (3.10.6)

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4502)

## [v0.4](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.4/images/sha256-ca5ba4aab952e6afb3150865b33b03846cf38d1b512fbae575d3f54f7d38a829?context=explore)

Compressed image size: 831.25 MB

This release only includes one major change on top of [release v0.3](release-notes#v03):

- Force 100% available VRAM size for compute runtime.

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4502)

For A770, now VRAM size increases from 13005 to 16256 MB.

```txt
v0.3 or earilier >>>> Torch detected GPU: Intel(R) Graphics [0x56a0] VRAM 13005 Compute Units 512
v0.4 >>>>>>>>>>>>>>>> Torch detected GPU: Intel(R) Graphics [0x56a0] VRAM 16256 Compute Units 512
```

## [v0.3](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.3/images/sha256-accb961e63a14b92567e7c594ad5222fd4592a40b8e3c5a76310a70257b1f00e?context=explore)

Compressed image size: 831.25 MB

Important notes:

* The severe VRAM leak problem of IPEX in WSL2 (https://github.com/Nuullll/ipex-sd-docker-for-arc-gpu/issues/8, https://github.com/intel/intel-extension-for-pytorch/issues/388) has been worked around in SD.Next https://github.com/vladmandic/automatic/commit/c3a4293f2227fe77b9ea908c99a1bda2aef43175. **If you are using WSL, please be sure to upgrade your SD.Next codebase by following [this instruction](getting-started.md#upgrade-sdnext-source-code).**

Major changes:

* Enable TCMalloc to mitigate the RAM leak problem in WSL.
* Mount `/root/.cache/huggingface` to avoid re-downloading huggingface models.

The image (Ubuntu 22.04 based) includes:

- Intel oneAPI DPC++ runtime libs (2023.1) (Note: compiler executables are not included)
- Intel oneAPI MKL runtime libs (2023.1)
- Intel oneAPI compiler common tool sycl-ls (2023.1)
- Intel Graphics driver (1.3.26241.21-647~22.04)
- Basic python environment (3.10.6)

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4502)

## [v0.2](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.2/images/sha256-58f7c7ae5b837b427623472a23582c1b4ecbd49460d245ddcb533e721cb396db?context=explore)

Compressed image size: 827.1 MB

Major changes:
- Removed IPEX and python packages required by SD.Next from the image to reduce size.
- Removed other unnecessary compiler tools (except sycl-ls) to reduce size.
- Uplift Intel Graphics driver to 1.3.26241.21-647~22.04.

The image (Ubuntu 22.04 based) includes:

- Intel oneAPI DPC++ runtime libs (2023.1) (Note: compiler executables are not included)
- Intel oneAPI MKL runtime libs (2023.1)
- Intel oneAPI compiler common tool sycl-ls (2023.1)
- Intel Graphics driver (1.3.26241.21-647~22.04)
- Basic python environment (3.10.6)

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4382)

## [v0.1](https://hub.docker.com/layers/nuullll/ipex-arc-sd/v0.1/images/sha256-5c00e46920a396a2b1c69e5ad24218883ba205afe6d59ce153f12f684ef2c006)

Compressed image size: 2.11 GB

Initial release. The image (Ubuntu 22.04 based) includes:

- Intel oneAPI DPC++ runtime libs (2023.1) (Note: compiler executables are not included)
- Intel oneAPI MKL runtime libs (2023.1)
- Intel oneAPI compiler common tool sycl-ls (2023.1)
- Intel Graphics driver (1.3.25593.18-601~22.04)
- Basic python environment (3.10.6)
- IPEX (1.13.120+xpu) and python packages required by SD.Next requirements.txt

Tested on Windows 11 22H2 22621.1848 with i9-13900 + Arc A770 (Windows driver: 31.0.101.4382)


scripts/package.ps1
# upgrade SD.Next
docker run -it `
--device /dev/dxg `
-v /usr/lib/wsl:/usr/lib/wsl `
-v $home\docker-mount\sd-webui-package:/sd-webui `
-v deps:/deps `
-v huggingface:/root/.cache/huggingface `
-p 7866:7860 `
--name sd-server-package-upgrade `
ipex-arc-sd --upgrade

# Copy localizations folder until https://github.com/vladmandic/automatic/pull/1783 is merged
Copy-Item -Path "$home\docker-mount\localizations" -Destination "$home\docker-mount\sd-webui-package\localizations" -Recurse

# prepare packaging container
docker run -it `
--device /dev/dxg `
-v /usr/lib/wsl:/usr/lib/wsl `
-v $home\docker-mount\sd-webui-package:/sd-webui `
-v deps:/deps `
-v huggingface:/root/.cache/huggingface `
-p 7866:7860 `
--name sd-server-package `
ipex-arc-sd

# export image
docker save --output $home\projects\arc-sd-all-in-one\image.tar nuullll/ipex-arc-sd:latest

# export volumes
docker pull ubuntu
docker run --rm --volumes-from sd-server-package -v $home\projects\arc-sd-all-in-one:/backup ubuntu tar cvf /backup/volume-deps.tar /deps
docker run --rm --volumes-from sd-server-package -v $home\projects\arc-sd-all-in-one:/backup ubuntu tar cvf /backup/volume-huggingface.tar /root/.cache/huggingface

# package Web UI folder
# exclude sensitive/large files
# sdnext.log
Remove-Item $home\docker-mount\sd-webui-package\sdnext.log
# outputs
Remove-Item $home\docker-mount\sd-webui-package\outputs\* -Recurse
# ControlNet
Move-Item $home\docker-mount\sd-webui-package\models\ControlNet\* $home\docker-mount\sd-webui-package-exclude\models\ControlNet -Force
# ControlNet annotator
Move-Item $home\docker-mount\sd-webui-package\extensions-builtin\sd-webui-controlnet\annotator\downloads\* $home\docker-mount\sd-webui-package-exclude\extensions-builtin\sd-webui-controlnet\annotator\downloads -Force
# SD models
Move-Item $home\docker-mount\sd-webui-package\models\Stable-diffusion\* $home\docker-mount\sd-webui-package-exclude\models\Stable-diffusion -Force
# embeddings
Move-Item $home\docker-mount\sd-webui-package\models\embeddings\* $home\docker-mount\sd-webui-package-exclude\models\embeddings -Force
# Lora
Move-Item $home\docker-mount\sd-webui-package\models\Lora\* $home\docker-mount\sd-webui-package-exclude\models\Lora -Force
# LyCORIS
Move-Item $home\docker-mount\sd-webui-package\models\LyCORIS\* $home\docker-mount\sd-webui-package-exclude\models\LyCORIS -Force

# Copy Web UI folder
Copy-Item $home\docker-mount\sd-webui-package\* -Destination $home\projects\arc-sd-all-in-one\webui -Recurse

# copy install script
Copy-Item $home\projects\ipex-sd-docker-for-arc-gpu\scripts\install.bat -Destination $home\projects\arc-sd-all-in-one

## Package
# Compress-Archive not working.. Use 7-zip manually.


scripts/package.py
import argparse
import logging
import os
from pathlib import Path
import requests
import shutil
from subprocess import Popen, PIPE

logging.basicConfig(format='%(asctime)s %(message)s', level=logging.DEBUG)

parser = argparse.ArgumentParser(description='Packaing script for Arc-SDNext-Installer')
parser.add_argument('output_dir', help='Directory to store the packaged artifacts')
parser.add_argument('--version', type=str, required=True, help='Version of packaged installer: <major>.<minor>.<patch>')
parser.add_argument('--stage', nargs='*', help='Packaging stages: image | volume | webui | strip | all')
parser.add_argument('--container', type=str, default='sd-server-package', help='The docker contaier name for packaging')
parser.add_argument('--webui-dir', help='The web ui directory')
parser.add_argument('--webui-dir-with-models', help='The opinionated web ui directory where models are available')

args = parser.parse_args()

MANUAL_CHECKS = [
    'localization file',
    'UI vae',
    'UI extra networks inline',
    'UI theme',
    'image process exact steps',
    'remove local wheels',
    'samplers',
    'webui-user.sh',
]

def run(cmd):
    with Popen(cmd, stdout=PIPE, stderr=PIPE) as p:
        o = p.stdout.read().decode('utf-8').strip()
        e = p.stderr.read().decode('utf-8').strip()
        if o:
            logging.info(o)
        if e:
            logging.error(e)
    
    if p.returncode != 0:
        raise Exception(f"Failed to run {' '.join(cmd)}")

def export_image():
    output_file = os.path.join(args.output_dir, 'image.tar')
    image_name = 'nuullll/ipex-arc-sd:latest'
    logging.info(f'Exporting image {image_name} to {output_file}')

    run(['docker', 'save', '--output', output_file, image_name])

def export_volumes():
    logging.info(f'Exporting volumes')
    run(['docker', 'pull', 'ubuntu'])
    run(['docker', 'run', '--rm', '-v', 'deps:/deps', '-v', f'{args.output_dir}:/backup', 'ubuntu', 'tar', 'cvf', '/backup/volume-deps.tar', '/deps'])
    run(['docker', 'run', '--rm', '-v', 'huggingface:/root/.cache/huggingface', '-v', f'{args.output_dir}:/backup', 'ubuntu', 'tar', 'cvf', '/backup/volume-huggingface.tar', '/root/.cache/huggingface'])

def launch_fresh_webui():
    logging.info(f'Setting up a fresh Web UI container (temporary)')
    run(['docker', 'run', '-d',
         '--device', '/dev/dxg',
         '-v', '/usr/lib/wsl:/usr/lib/wsl',
         '-v', f'{args.webui_dir}:/sd-webui',
         '-v', 'deps:/deps',
         '-v', 'huggingface:/root/.cache/huggingface',
         '-p', '7866:7860',
         '-e', 'TORCH_COMMAND=intel_extension_for_pytorch-2.0.110+xpu-cp310-cp310-linux_x86_64.whl torch-2.0.1a0+cxx11.abi-cp310-cp310-linux_x86_64.whl torchvision-0.15.2a0+cxx11.abi-cp310-cp310-linux_x86_64.whl',
         '--rm',
         'nuullll/ipex-arc-sd:latest',
         '--upgrade', '--no-download'])
    input('Press Enter to continue when Web UI is ready ...')
    
def copy_fresh_webui():
    dst = os.path.join(args.output_dir, 'webui')
    logging.info(f'Copying {args.webui_dir} to {dst}')
    shutil.copytree(args.webui_dir, dst, ignore=shutil.ignore_patterns('sdnext.log', 'outputs/**/*'))

def install_plugins_to(webui_dir):
    modified_dirs = []
    def new_dir(dir_name):
        full_path = os.path.join(webui_dir, dir_name)
        os.makedirs(full_path, exist_ok=True)
        modified_dirs.append(full_path)
        return full_path
    # bilingual localizations
    ## download localization file
    r = requests.get('https://gist.githubusercontent.com/journey-ad/d98ed173321658be6e51f752d6e6163c/raw/aa162c81d9d7d7b6efdffa84ccbfe867be6b711d/I18N_sd-webui-zh_CN.json')
    locale_dir = new_dir('localizations')
    with open(os.path.join(locale_dir, 'I18N_sd-webui-zh_CN.json'), 'wb') as f:
        f.write(r.content)

    ext_dir = new_dir('extensions')
    def add_extension(git_url):
        ext_name = git_url.split('/')[-1][:-4]
        logging.info(f"Cloning extension {ext_name} from {git_url}")
        run(['git', 'clone', git_url, os.path.join(ext_dir, ext_name)])
    # custom extensions
    add_extension('https://github.com/journey-ad/sd-webui-bilingual-localization.git')
    add_extension('https://github.com/DominikDoom/a1111-sd-webui-tagcomplete.git')
    add_extension('https://github.com/Physton/sd-webui-prompt-all-in-one.git')
    add_extension('https://github.com/Coyote-A/ultimate-upscale-for-automatic1111.git')
    add_extension('https://github.com/Bing-su/adetailer.git')

    return modified_dirs

def copy_plugins():
    par_dir = os.path.dirname(args.webui_dir)
    basename = os.path.basename(args.webui_dir)
    ws = os.path.join(par_dir, basename + '-plugins')
    if not os.path.exists(ws):
        logging.info(f'Making a copy of fresh Web UI for plugin installation...')
        fresh_webui = os.path.join(args.output_dir, 'webui')
        shutil.copytree(fresh_webui, ws)
    
    modified_dirs = install_plugins_to(ws)
    for d in modified_dirs:
        rel_path = os.path.relpath(d, ws)
        shutil.copytree(d, os.path.join(args.output_dir, 'webui-plugins', rel_path))

def get_dir_size(dir):
    return sum(f.stat().st_size for f in Path(dir).glob('**/*') if f.is_file()) / (1024*1024*1024)

def copy_models():
    # models
    model_dir = os.path.join(args.webui_dir_with_models, 'models')
    exclude_dirs = ['Lora', 'LyCORIS', 'Stable-diffusion']
    basic_model_output = os.path.join(args.output_dir, 'webui-models')
    shutil.copytree(model_dir, os.path.join(basic_model_output, 'models'), ignore=shutil.ignore_patterns(*exclude_dirs))
    logging.info(f"Copied basic models: {get_dir_size(basic_model_output)} GB")

    # extra large models
    extra_model_output = os.path.join(args.output_dir, 'webui-extra-models')
    ## one SD model
    model_name = 'majicmixRealistic_betterV2V25.safetensors'
    dst = os.path.join(extra_model_output, 'models', 'Stable-diffusion', model_name)
    os.makedirs(os.path.dirname(dst), exist_ok=True)
    shutil.copy2(os.path.join(model_dir, 'Stable-diffusion', model_name), dst)
    ## ControlNet preprocessors
    annotator_path = ['extensions-builtin', 'sd-webui-controlnet', 'annotator', 'downloads']
    cn_annotator_dir = os.path.join(args.webui_dir_with_models, *annotator_path)
    shutil.copytree(cn_annotator_dir, os.path.join(extra_model_output, *annotator_path))
    ## ControlNet models
    cn_model_path = ['extensions-builtin', 'sd-webui-controlnet', 'models']
    shutil.copytree(os.path.join(args.webui_dir_with_models, *cn_model_path), os.path.join(extra_model_output, *cn_model_path))
    logging.info(f"Copied extra models: {get_dir_size(extra_model_output)} GB")

def manual_check():
    par_dir = os.path.dirname(args.webui_dir)
    basename = os.path.basename(args.webui_dir)
    ws = os.path.join(par_dir, basename + '-plugins')
    os.remove(os.path.join(ws, 'sdnext.log'))
    run(['docker', 'run', '-d',
         '--device', '/dev/dxg',
         '-v', '/usr/lib/wsl:/usr/lib/wsl',
         '-v', f'{ws}:/sd-webui',
         '-v', 'deps:/deps',
         '-v', 'huggingface:/root/.cache/huggingface',
         '-p', '7866:7860',
         '--rm',
         'nuullll/ipex-arc-sd:latest',
         '--insecure', '--no-download'])
    for item in MANUAL_CHECKS:
        input(f"Check item {item}")

    input("All done! Save edits then press ENTER")
    
    dst_config = os.path.join(args.output_dir, 'webui', 'config.json')
    if os.path.exists(dst_config):
        os.remove(dst_config)
    shutil.copy2(os.path.join(ws, 'config.json'), dst_config)
    logging.info(f"{dst_config} updated!")

if __name__ == '__main__':
    major, minor, patch = args.version.split('.')
    os.makedirs(args.output_dir, exist_ok=True)

    stages = args.stage
    if 'image' in stages or 'all' in stages:
        export_image()

    if 'setup-fresh' in stages or 'all' in stages:
        launch_fresh_webui()

    if 'webui' in stages or 'all' in stages:
        copy_fresh_webui()

    if 'webui-plugins' in stages or 'all' in stages:
        copy_plugins()

    if 'webui-models' in stages or 'all' in stages:
        copy_models()

    if 'manual-check' in stages or 'all' in stages:
        manual_check()

    if 'volume' in stages or 'all' in stages:
        export_volumes()
